jib-core/src/main/java/com/google/cloud/tools/jib/api/JavaContainerBuilder.java

```diff
@@ -0,0 +1,305 @@
+    return from(RegistryImage.named(baseImageReference));
+  }
+
+  /**
+   * Creates a new {@link JavaContainerBuilder} with the specified base image reference.
+   *
+   * @param baseImageReference the base image reference
+   * @return a new {@link JavaContainerBuilder}
+   */
+  public static JavaContainerBuilder from(ImageReference baseImageReference) {
+    return from(RegistryImage.named(baseImageReference));
+  }
+
+  /**
+   * Creates a new {@link JavaContainerBuilder} with the specified base image.
+   *
+   * @param registryImage the {@link RegistryImage} that defines base container registry and
+   *     credentials
+   * @return a new {@link JavaContainerBuilder}
+   */
+  public static JavaContainerBuilder from(RegistryImage registryImage) {
+    return new JavaContainerBuilder(Jib.from(registryImage));
+  }
+
+  private final JibContainerBuilder jibContainerBuilder;
+  private final JavaLayerConfigurations.Builder layerConfigurationsBuilder =
+      JavaLayerConfigurations.builder();
+  private final List<String> jvmFlags = new ArrayList<>();
+  private final LinkedHashSet<String> classpath = new LinkedHashSet<>(4);
+
+  @Nullable private String mainClass;
+
+  private JavaContainerBuilder(JibContainerBuilder jibContainerBuilder) {
+    this.jibContainerBuilder = jibContainerBuilder;
+  }
+
+  /**
+   * Adds dependency JARs to the image.
+   *
+   * @param dependencyFiles the list of dependency JARs to add to the image
+   * @return this
+   * @throws NoSuchFileException if adding the layer fails
+   */
+  public JavaContainerBuilder addDependencies(List<Path> dependencyFiles)
+      throws NoSuchFileException {
+    // Make sure all files exist before adding any
+    for (Path file : dependencyFiles) {
+      if (!Files.exists(file)) {
+        throw new NoSuchFileException(file.toString());
+      }
+    }
+
+    for (Path file : dependencyFiles) {
+      layerConfigurationsBuilder.addFile(
+          file.getFileName().toString().contains("SNAPSHOT")
+              ? LayerType.SNAPSHOT_DEPENDENCIES
+              : LayerType.DEPENDENCIES,
+          file,
+          DEPENDENCIES_PATH.resolve(file.getFileName()));
+    }
+    classpath.add(DEPENDENCIES_CLASSPATH.toString());
+    return this;
+  }
+
+  /**
+   * Adds dependency JARs to the image.
+   *
+   * @param dependencyFiles the list of dependency JARs to add to the image
+   * @return this
+   * @throws NoSuchFileException if adding the layer fails
+   */
+  public JavaContainerBuilder addDependencies(Path... dependencyFiles) throws NoSuchFileException {
+    return addDependencies(Arrays.asList(dependencyFiles));
+  }
+
+  /**
+   * Adds the contents of a resources directory to the image.
+   *
+   * @param resourceFilesDirectory the directory containing the project's resources
+   * @return this
+   * @throws IOException if adding the layer fails
+   */
+  public JavaContainerBuilder addResources(Path resourceFilesDirectory) throws IOException {
```

Comment:
This may work for many jib-core uses, but it is a bit limited in that, e.g., our plugins won't be able to use this public API. For example, class files and resource files will be mixed under `target/classes`.

I'm not saying this has to be addressed in this PR, but keep in mind that if the API is like this, our plugins cannot be the user of this API. But then, I guess you don't intend for our plugins to use this API, since it doesn't have the ability to change the app root or put some files in the app root?

Reply:
Maybe we should add overloads with a second parameter that's a path filter? I think for app root we might want to keep that out of this API and just have the user use the `JibContainerBuilder` API to work with that.